paddle/framework/net_design.md

```diff
@@ -0,0 +1,159 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators in a network, 
+users can build a real network from a `NetDef` in protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetworkBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// The minimum a network should be implemented.
+class NetworkBase {
+ public:
+  // `def` is a proto message that describe the structure of a network.
+  NetworkBase(const NetDef& def);
+
+  // run all the operators and return success(true) or not, all the
+  // variables are located in `scope`.
+  virtual bool Run(Scope* scope) = 0;
```

Comment:
There is a discussion in [here](https://github.com/PaddlePaddle/Paddle/pull/2548#issuecomment-310555355).  Should our `Operator` be `stateless`? If our `Operator` is `stateless`, then `NetBase::Run()` should be `const`.